File: coding_int/chapter_1/1_4_palindrome_permutation.py
# scenarios:
# basically count letters if total character count is even then all counts must be even
# else allow one set of odd
# lower and strip spaces then character count?

# odd / even is by modulo 2

def isPallindrome(striga):

    transform = striga.lower()
    transform = transform.replace(" ", "")

    holder = {}
    length_string = 0

    for char in transform:
        if char in holder.keys():
            holder[char] += 1
        else:
            holder[char] = 1

        length_string += 1
    
    if (length_string % 2 == 0):

        for key in holder.keys():
            if (holder[key] % 2 != 0):
                return False
        
        return True
    
    elif (length_string % 2 != 0):
        odd = 0

        for key in holder.keys():
            if( holder[key] % 2 != 0):
                odd += 1
                if (odd > 1):
                    return False
        
        return True
    
    raise ValueError("our logic failed")

if __name__ == "__main__":
    
    a = "eat food" # should be false
    b = "abba" # should be true
    c = "eateatfood" # should be false
    d = "eat eat food" # should be false
    e = "eat tae dood" # should be true

    list_tests = [a,b,c,d,e]

    for item in list_tests:
        print(isPallindrome(item))

File: coding_int/chapter_1/1_4_palindrome_permutation.cpp
#include<iostream>
#include<string>
#include<unordered_map>

bool isPallindrome(std::string abba) {

    std::unordered_map<char, int> holder;

    for (std::string::size_type i = 0; i < abba.length(); ++i) {
        ++holder[abba[i]];
    }

    if (abba.length() % 2 == 0) {
        for (auto it : holder) {
            if (it.second % 2 != 0) {
                return false;
            }
        }

        return true;

    } else if (abba.length() % 2 != 0) {
        int odd = 0;
        for (auto it : holder) {
            if (it.second % 2 != 0) {
                ++odd;

                if (odd > 1) {
                    return false;
                }
            }
        }
        return true;
    }

    return false;
}

int main() {

    std::string a = "eat food"; // should be false
    std::string b = "abba"; // should be true
    std::string c = "eateatfood"; // should be false
    std::string d = "eat eat food"; // should be false
    std::string e = "eat tae dood"; // should be true

    std::cout << "case is: " << a << " result is: " << isPallindrome(a) << std::endl;
    std::cout << "case is: " << b << " result is: " << isPallindrome(b) << std::endl;
    std::cout << "case is: " << c << " result is: " << isPallindrome(c) << std::endl;
    std::cout << "case is: " << d << " result is: " << isPallindrome(d) << std::endl;
    std::cout << "case is: " << e << " result is: " << isPallindrome(e) << std::endl;

    return 1;
}
